insights/v1/types_insights.go

```diff
@@ -0,0 +1,440 @@
+	// +kubebuilder:validation:XValidation:rule=`isURL(self) && url(self).getScheme() == "https"`,message=`URI must be a valid HTTPS URL (e.g., https://example.com)`
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=2048
+	// +optional
+	URI string `json:"uri,omitempty,omitzero"`
+}
+
+// healthCheck represents an Insights health check attributes.
+type HealthCheck struct {
+	// description is required field that provides basic description of the healtcheck.
+	// It must contain at least 10 characters and may not exceed 2048 characters.
+	// +kubebuilder:validation:MinLength=10
+	// +kubebuilder:validation:MaxLength=2048
+	// +required
+	Description string `json:"description,omitempty"`
+	// totalRisk is the required field of the healthcheck.
+	// It is indicator of the total risk posed by the detected issue; combination of impact and likelihood.
+	// Allowed values are Low, Moderate, Important and Critical.
+	// The value represents the severity of the issue.
+	// +required
+	TotalRisk TotalRisk `json:"totalRisk,omitempty"`
+	// advisorURI is required field that provides the URL link to the Insights Advisor.
+	// The link must be a valid HTTPS URL and the maximum length is 2048 characters.
+	// +kubebuilder:validation:XValidation:rule=`isURL(self) && url(self).getScheme() == "https"`,message=`advisorURI must be a valid HTTPS URL (e.g., https://example.com)`
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=2048
+	// +required
+	AdvisorURI string `json:"advisorURI,omitempty"`
+}
+
+// totalRisk defines the valid totalRisk values.
+// +kubebuilder:validation:Enum=Low;Moderate;Important;Critical
+type TotalRisk string
+
+const (
+	TotalRiskLow       TotalRisk = "Low"
+	TotalRiskModerate  TotalRisk = "Moderate"
+	TotalRiskImportant TotalRisk = "Important"
+	TotalRiskCritical  TotalRisk = "Critical"
+)
+
+// ObjectReference contains enough information to let you inspect or modify the referred object.
+type ObjectReference struct {
+	// group is required field that specifies the API Group of the Resource.
+	// Enter empty string for the core group.
+	// This value is empty or it should follow the DNS1123 subdomain format.
+	// It must be at most 253 characters in length, and must consist only of lower case alphanumeric characters, '-' and '.', and must start with an alphabetic character and end with an alphanumeric character.
+	// Example: "", "apps", "build.openshift.io", etc.
+	// +kubebuilder:validation:XValidation:rule="self.size() == 0 || !format.dns1123Subdomain().validate(self).hasValue()",message="a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start with an alphabetic character and end with an alphanumeric character."
+	// +kubebuilder:validation:MinLength=0
+	// +kubebuilder:validation:MaxLength=253
+	// +required
+	Group string `json:"group,omitempty"`
```

Comment:
Linter is forcing to set this field as a pointer, because the value can be set to `""` which in this case represents the core group. Do I need the pointer when this field is required? @JoelSpeed @everettraven

Reply:
Yes, this should be a pointer because there is a difference between `""` and unset. Without a pointer, and with omitempty, this field wouldn't round trip when set to the `""` value. Structured clients must be forced to make a choice for all fields, even when `""` is their choice, as such, this should be a pointer so that structured clients can set the value deliberately to `""`